src/sage/rings/species.py

```diff
@@ -0,0 +1,1331 @@
+            raise NotImplementedError("Cannot cancel in this monoid")
+
+        def _mul_(self, other):
+            r"""
+            Multiply ``self`` by ``other``.
+            """
+            res = super()._mul_(other)
+            elm = self.parent()._cache_get(res)
+            if elm._group is None:
+                elm._elmmul(self, other)
+                elm._canonicalize()
+            return elm
+
+        def _elmexp(self, other, n):
+            r"""
+            Populate the group info of ``self`` by exponentiating
+            the group of ``other``.
+            """
+            if other._group is None:
+                other._group_constructor()
+            temp = self.parent().one()
+            while n > 0:
+                if n % 2 == 1:
+                    temp = temp * other
+                other = other * other
+                n //= 2
+            self._assign_group_info(temp)
+
+        def __pow__(self, n):
+            r"""
+            Raise ``self`` to the power of ``n``.
+            """
+            res = super().__pow__(n)
+            elm = self.parent()._cache_get(res)
+            if elm._group is None:
+                elm._elmexp(self, n)
+                elm._canonicalize()
+            return elm
+
+        def _element_key(self):
+            r"""
+            Return the cache lookup key for ``self``.
+            """
+            return self
+
+        @cached_method
+        def _canonicalize(self):
+            r"""
+            Canonicalize this molecular species by sorting the orbits by
+            length and making them consecutive.
+
+            EXAMPLES::
+
+                sage: P = PolynomialSpecies(ZZ, "X, Y")
+                sage: G = PermutationGroup([[(1,2),(3,4),(5,6),(7,8,9,10)]]); G
+                Permutation Group with generators [(1,2)(3,4)(5,6)(7,8,9,10)]
+                sage: H = PermutationGroup([[(1,2,3,4),(5,6),(7,8),(9,10)]]); H
+                Permutation Group with generators [(1,2,3,4)(5,6)(7,8)(9,10)]
+                sage: A = P(G, {1: [1,2,3,4], 2: [5,6,7,8,9,10]})
+                sage: A.support()[0]._dompart
+                ((5, 6, 7, 8), (9, 10, 1, 2, 3, 4))
+                sage: C = P(G, {1: [1,2,5,6], 2: [3,4,7,8,9,10]})
+                sage: C.support()[0]._dompart
+                ((5, 6, 7, 8), (9, 10, 1, 2, 3, 4))
+            """
+            if self._group is None or self._group == SymmetricGroup(0):
+                return
+            sorted_orbits = sorted([sorted(orbit) for orbit in self._group.orbits()], key=len, reverse=True)
+            pi = PermutationGroupElement(list(chain.from_iterable(sorted_orbits))).inverse()
+            self._group = PermutationGroup(gap_group=libgap.ConjugateGroup(self._group, pi))
+            self._dompart = tuple(tuple(pi(k) for k in v) for v in self._dompart)
+
+        def grade(self):
+            r"""
+            Return the grade of ``self``.
+            """
+            return self._tc
```

Comment:
Shouldn't these be multi-graded, just as `AtomicSpecies`?

Reply:
oh no
yes that makes sense...